Mismatched type error: type of crate instead of std is displayed

On nightly rustc (cargo) I get misleading expected type errors. If a crate has the same(?) type it is displayed instead of the std type.

I tried this code:
```rust
extern crate bitflags;
fn test(r: Result<Option<()>, &'static str>) { }
fn main() { test(Ok(())); }
```

I expected to see this happen: (stable rust)
```
$ rustup default stable
$ cargo build
   Compiling rust-sample v0.1.0 (file:///[...]/rust-sample)
error[E0308]: mismatched types
 --> src/main.rs:3:21
  |
3 | fn main() { test(Ok(())); }
  |                     ^^
  |                     |
  |                     expected enum `std::option::Option`, found ()
  |                     help: try using a variant of the expected type: `Some(())`
  |
  = note: expected type `std::option::Option<()>`
             found type `()`
```

Instead, this happened: (nightly)
```
$ rustup default nightly
$ cargo build
   Compiling rust-sample v0.1.0 (file:///[...]/rust-sample)
error[E0308]: mismatched types
 --> src/main.rs:3:21
  |
3 | fn main() { test(Ok(())); }
  |                     ^^
  |                     |
  |                     expected enum `bitflags::<unnamed>::option::Option`, found ()
  |                     help: try using a variant of the expected type: `bitflags::<unnamed>::prelude::v1::Some(())`
  |
  = note: expected type `bitflags::<unnamed>::option::Option<()>`
             found type `()`
```

## Meta

`rustc --version --verbose`:
```
rustc 1.23.0-nightly (5041b3bb3 2017-11-19)
binary: rustc
commit-hash: 5041b3bb3d953a14f32b15d1e41341c629acae12
commit-date: 2017-11-19
host: x86_64-unknown-linux-gnu
release: 1.23.0-nightly
LLVM version: 4.0

```

--- a/src/librustc_metadata/cstore_impl.rs
+++ b/src/librustc_metadata/cstore_impl.rs
@@ -294,13 +294,31 @@ pub fn provide<'tcx>(providers: &mut Providers<'tcx>) {
             assert_eq!(cnum, LOCAL_CRATE);
             let mut visible_parent_map: DefIdMap<DefId> = DefIdMap();
 
+            // Issue 46112: We want the map to prefer the shortest
+            // paths when reporting the path to an item. Therefore we
+            // build up the map via a breadth-first search (BFS),
+            // which naturally yields minimal-length paths.
+            //
+            // Note that it needs to be a BFS over the whole forest of
+            // crates, not just each individual crate; otherwise you
+            // only get paths that are locally minimal with respect to
+            // whatever crate we happened to encounter first in this
+            // traversal, but not globally minimal across all crates.
+            let bfs_queue = &mut VecDeque::new();
             for &cnum in tcx.crates().iter() {
                 // Ignore crates without a corresponding local `extern crate` item.
                 if tcx.missing_extern_crate_item(cnum) {
                     continue
                 }
 
-                let bfs_queue = &mut VecDeque::new();
+                bfs_queue.push_back(DefId {
+                    krate: cnum,
+                    index: CRATE_DEF_INDEX
+                });
+            }
+
+            // (restrict scope of mutable-borrow of `visible_parent_map`)
+            {
                 let visible_parent_map = &mut visible_parent_map;
                 let mut add_child = |bfs_queue: &mut VecDeque<_>,
                                      child: &def::Export,
@@ -326,10 +344,6 @@ pub fn provide<'tcx>(providers: &mut Providers<'tcx>) {
                     }
                 };
 
-                bfs_queue.push_back(DefId {
-                    krate: cnum,
-                    index: CRATE_DEF_INDEX
-                });
                 while let Some(def) = bfs_queue.pop_front() {
                     for child in tcx.item_children(def).iter() {
                         add_child(bfs_queue, child, def);

--- a/src/test/ui/auxiliary/xcrate_issue_46112_rexport_core.rs
+++ b/src/test/ui/auxiliary/xcrate_issue_46112_rexport_core.rs
@@ -1,0 +1,13 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![crate_type="lib"]
+
+pub extern crate core;

--- a/src/test/ui/issue-46112.rs
+++ b/src/test/ui/issue-46112.rs
@@ -1,0 +1,20 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Issue 46112: An extern crate pub reexporting libcore was causing
+// paths rooted from `std` to be misrendered in the diagnostic output.
+
+// ignore-windows
+// aux-build:xcrate_issue_46112_rexport_core.rs
+
+extern crate xcrate_issue_46112_rexport_core;
+fn test(r: Result<Option<()>, &'static str>) { }
+fn main() { test(Ok(())); }
+//~^ mismatched types

--- a/src/test/ui/issue-46112.stderr
+++ b/src/test/ui/issue-46112.stderr
@@ -1,0 +1,14 @@
+error[E0308]: mismatched types
+  --> $DIR/issue-46112.rs:19:21
+   |
+19 | fn main() { test(Ok(())); }
+   |                     ^^
+   |                     |
+   |                     expected enum `std::option::Option`, found ()
+   |                     help: try using a variant of the expected type: `Some(())`
+   |
+   = note: expected type `std::option::Option<()>`
+              found type `()`
+
+error: aborting due to previous error
+

